File: 228__Summary_Ranges.py
class Solution(object):
    def summaryRanges(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        if len(nums) == 0:
            return []
        start, end = 0, 0
        result = []
        for i in range(1, len(nums)):
            if nums[i] == nums[end] + 1:
                end += 1
            else:
                if start == end:
                    result.append(str(nums[start]))
                else:
                    result.append(str(nums[start]) + "->" + str(nums[end]))
                start, end = i, i

        if start == end:
            result.append(str(nums[start]))
        else:
            result.append(str(nums[start]) + "->" + str(nums[end]))

        return result

# Faster using while loop

class Solution(object):
    def summaryRanges(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        if len(nums) == 0:
            return []

        left, right = 0, 0
        result = []
        while right < len(nums):
            while right + 1 < len(nums) and nums[right] + 1 == nums[right + 1]:
                right += 1
            if left < right:
                result.append(str(nums[left]) + "->" + str(nums[right]))
            else:
                result.append(str(nums[left]))
            left, right = right + 1, right + 1

        return result

File: Java/228__Summary_Ranges.java
class Solution {
    public List<String> summaryRanges(int[] nums) {

        List<String> result = new ArrayList<>();
        if (nums.length == 0) {
            return result;
        }
        int start = Integer.MAX_VALUE, end = Integer.MIN_VALUE;
        for (int i = 0; i < nums.length; i++) {
            if (start == Integer.MAX_VALUE) {
                start = nums[i];
            } else {
                if (end == Integer.MIN_VALUE) {
                    if (nums[i] == start + 1) {
                        end = nums[i];
                    } else {
                        result.add(Integer.toString(start));
                        start = nums[i];
                    }
                } else {
                    if (nums[i] == end + 1) {
                        end = nums[i];
                    } else {
                        result.add(Integer.toString(start) + "->" + Integer.toString(end));
                        start = nums[i];
                        end = Integer.MIN_VALUE;
                    }
                }
            }
        }
        if (end == Integer.MIN_VALUE) {
            result.add(Integer.toString(start));
        } else {
            result.add(Integer.toString(start) + "->" + Integer.toString(end));
        }
        return result;

    }
}
